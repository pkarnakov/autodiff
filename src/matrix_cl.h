#pragma once

#include <algorithm>
#include <cmath>
#include <iomanip>
#include <iosfwd>
#include <numeric>
#include <vector>

#include "macros.h"
#include "matrix.h"
#include "opencl.h"

template <class T>
class MatrixCL {
 public:
  using CL = OpenCL;
  using MSize = OpenCL::MSize;
  using MInt = OpenCL::MInt;
  template <class U>
  friend class Matrix;

  class Entry {
   public:
    Entry(MatrixCL& matr, int i, int j) : matr_(matr), i_(i), j_(j) {}
    Entry& operator=(T value) {
      matr_.write(i_, j_, value);
      return *this;
    }
    operator T() const {
      return matr_.read(i_, j_);
    }

   private:
    MatrixCL& matr_;
    int i_;
    int j_;
  };

  // Constructor.
  MatrixCL() = default;
  MatrixCL(MSize nw, CL* cl_)
      : cl(cl_), nw_(nw), data_(cl->context(), nw_[0] * nw_[1]) {}
  MatrixCL(MSize nw, CL& cl_) : MatrixCL(nw, &cl_) {}
  MatrixCL(size_t nrow, size_t ncol, CL* cl_) : MatrixCL({nrow, ncol}, cl_) {}
  MatrixCL(size_t nrow, size_t ncol, CL& cl_) : MatrixCL({nrow, ncol}, &cl_) {}
  MatrixCL(MSize nw, T value, CL* cl_) : MatrixCL(nw, cl_) {
    this->fill(value);
  }
  MatrixCL(size_t nrow, size_t ncol, T value, CL* cl_)
      : MatrixCL({nrow, ncol}, value, cl_) {}
  MatrixCL(const MatrixCL& other) : MatrixCL(other.nw_, other.cl) {
    fassert(data_.handle);
    fassert(other.data_.handle);
    data_.EnqueueWriteBuffer(cl->queue(), other.data_);
  }
  MatrixCL(MatrixCL&&) = default;
  MatrixCL(const Matrix<T>& other, CL& cl_)
      : MatrixCL(other.nrow(), other.ncol(), cl_) {
    data_.EnqueueWrite(cl->queue(), other.data());
  }

  // Element access.
  T read(int i, int j) const {
    return cl->ReadAt<T>(nw_, data_, {i, j});
  }
  void write(int i, int j, T value) {
    return cl->WriteAt(nw_, data_, {i, j}, value);
  }
  T operator()(int i, int j) const {
    return read(i, j);
  }
  Entry operator()(int i, int j) {
    return Entry(*this, i, j);
  }
  size_t nrow() const {
    return nw_[0];
  }
  size_t ncol() const {
    return nw_[1];
  }
  size_t size() const {
    return nw_[0] * nw_[1];
  }

  // Assignment.
  MatrixCL& operator=(const MatrixCL& other) {
    fassert(other.cl, "Can't assign add an empty matrix");
    if (cl == nullptr) {
      cl = other.cl;
      nw_ = other.nw_;
      CL::Buffer<T> buf(cl->context(), nw_[0] * nw_[1]);
      data_.swap(buf);
    }
    fassert_equal(nw_[0], other.nw_[0]);
    fassert_equal(nw_[1], other.nw_[1]);
    fassert(data_.handle);
    fassert(other.data_.handle);
    data_.EnqueueWriteBuffer(cl->queue(), other.data_);
    return *this;
  }
  MatrixCL& operator+=(const MatrixCL& other) {
    fassert(other.cl, "Can't assign add an empty matrix");
    if (cl == nullptr) {
      cl = other.cl;
      nw_ = other.nw_;
      CL::Buffer<T> buf(cl->context(), nw_[0] * nw_[1]);
      data_.swap(buf);
    }
    fassert_equal(nw_[0], other.nw_[0]);
    fassert_equal(nw_[1], other.nw_[1]);
    cl->AssignAdd(nw_, data_, other.data_);
    return *this;
  }
  MatrixCL& operator-=(const MatrixCL& other) {
    fassert(other.cl, "Can't assign add an empty matrix");
    if (cl == nullptr) {
      cl = other.cl;
      nw_ = other.nw_;
      CL::Buffer<T> buf(cl->context(), nw_[0] * nw_[1]);
      data_.swap(buf);
    }
    fassert_equal(nw_[0], other.nw_[0]);
    fassert_equal(nw_[1], other.nw_[1]);
    cl->AssignSub(nw_, data_, other.data_);
    return *this;
  }
  void fill(T value) const {
    return cl->Fill(nw_, data_, value);
  }
  void clear() const {
    if (cl) {
      cl->Fill(nw_, data_, T(0));
    }
  }
  explicit operator Matrix<T>() const {
    Matrix<T> res(nw_[0], nw_[1]);
    data_.EnqueueRead(cl->queue(), res.data());
    return res;
  }

  // Member functions.
  MatrixCL operator-() const {
    MatrixCL res(nw_, cl);
    cl->Mul(nw_, data_, -1, res.data_);
    return res;
  }
  MatrixCL operator+(const MatrixCL& other) const {
    fassert_equal(cl, other.cl);
    fassert_equal(nw_[0], other.nw_[0]);
    fassert_equal(nw_[1], other.nw_[1]);
    MatrixCL res(nw_, cl);
    cl->Add(nw_, data_, other.data_, res.data_);
    return res;
  }
  MatrixCL operator-(const MatrixCL& other) const {
    fassert_equal(cl, other.cl);
    fassert_equal(nw_[0], other.nw_[0]);
    fassert_equal(nw_[1], other.nw_[1]);
    MatrixCL res(nw_, cl);
    cl->Sub(nw_, data_, other.data_, res.data_);
    return res;
  }
  MatrixCL operator*(const MatrixCL& other) const {
    fassert_equal(cl, other.cl);
    fassert_equal(nw_[0], other.nw_[0]);
    fassert_equal(nw_[1], other.nw_[1]);
    MatrixCL res(nw_, cl);
    cl->Mul(nw_, data_, other.data_, res.data_);
    return res;
  }
  MatrixCL operator/(const MatrixCL& other) const {
    fassert_equal(cl, other.cl);
    fassert_equal(nw_[0], other.nw_[0]);
    fassert_equal(nw_[1], other.nw_[1]);
    MatrixCL res(nw_, cl);
    cl->Div(nw_, data_, other.data_, res.data_);
    return res;
  }
  MatrixCL operator+(T a) const {
    MatrixCL res(nw_, cl);
    cl->Add(nw_, data_, a, res.data_);
    return res;
  }
  MatrixCL operator-(T a) const {
    MatrixCL res(nw_, cl);
    cl->Sub(nw_, data_, a, res.data_);
    return res;
  }
  MatrixCL operator*(T a) const {
    MatrixCL res(nw_, cl);
    cl->Mul(nw_, data_, a, res.data_);
    return res;
  }
  MatrixCL operator/(T a) const {
    MatrixCL res(nw_, cl);
    cl->Div(nw_, data_, a, res.data_);
    return res;
  }
  MatrixCL roll(int shift_row, int shift_col) const {
    MatrixCL res(nw_, cl);
    cl->Roll(nw_, data_, {shift_col, shift_row}, res.data_);
    return res;
  }
  MatrixCL restrict() const {
    fassert(nw_[0] % 2 == 0);
    fassert(nw_[1] % 2 == 0);
    MatrixCL res(nw_[0] / 2, nw_[1] / 2, cl);
    cl->Restrict(res.nw_, data_, res.data_);
    return res;
  }
  MatrixCL restrict_adjoint() const {
    MatrixCL res(nw_[0] * 2, nw_[1] * 2, cl);
    cl->RestrictAdjoint(nw_, data_, res.data_);
    return res;
  }
  MatrixCL interpolate() const {
    MatrixCL res(nw_[0] * 2, nw_[1] * 2, cl);
    cl->Interpolate(res.nw_, data_, res.data_);
    return res;
  }
  MatrixCL interpolate_adjoint() const {
    fassert(nw_[0] % 2 == 0);
    fassert(nw_[1] % 2 == 0);
    MatrixCL res({nw_[0] / 2, nw_[1] / 2}, cl);
    cl->InterpolateAdjoint(res.nw_, data_, res.data_);
    return res;
  }
  MatrixCL conv(const T& a, const T& axm, const T& axp, const T& aym,
                const T& ayp) const {
    MatrixCL res(nw_, cl);
    if (nw_[0] == 0 || nw_[1] == 0) {
      return res;
    }
    // Swapping the order of x and y.
    // TODO: Revise to make x the slow index everywhere.
    cl->Conv(nw_, data_, a, aym, ayp, axm, axp, res.data_);
    return res;
  }

  // Reduction.
  T sum() const {
    return cl->Sum(nw_, data_);
  }
  T mean() const {
    return sum() / size();
  }
  T dot(const MatrixCL& other) const {
    fassert_equal(cl, other.cl);
    return cl->Dot(nw_, data_, other.data_);
  }
  T max() const {
    return cl->Max(nw_, data_);
  }
  T min() const {
    return cl->Min(nw_, data_);
  }

  // Friend functions.
  friend MatrixCL sin(const MatrixCL& matr) {
    MatrixCL res(matr.nw_, matr.cl);
    matr.cl->Sin(matr.nw_, matr.data_, res.data_);
    return res;
  }
  friend MatrixCL cos(const MatrixCL& matr) {
    MatrixCL res(matr.nw_, matr.cl);
    matr.cl->Cos(matr.nw_, matr.data_, res.data_);
    return res;
  }
  friend MatrixCL exp(const MatrixCL& matr) {
    MatrixCL res(matr.nw_, matr.cl);
    matr.cl->Exp(matr.nw_, matr.data_, res.data_);
    return res;
  }
  friend MatrixCL log(const MatrixCL& matr) {
    MatrixCL res(matr.nw_, matr.cl);
    matr.cl->Log(matr.nw_, matr.data_, res.data_);
    return res;
  }
  friend MatrixCL sqr(const MatrixCL& matr) {
    MatrixCL res(matr.nw_, matr.cl);
    matr.cl->Sqr(matr.nw_, matr.data_, res.data_);
    return res;
  }
  friend MatrixCL sqrt(const MatrixCL& matr) {
    MatrixCL res(matr.nw_, matr.cl);
    matr.cl->Sqrt(matr.nw_, matr.data_, res.data_);
    return res;
  }
  friend T sum(const MatrixCL& matr) {
    return matr.sum();
  }
  friend T mean(const MatrixCL& matr) {
    return matr.mean();
  }
  friend T rms(const MatrixCL& matr) {
    using std::sqrt;
    return sqrt(sqr(matr).mean());
  }
  friend T dot(const MatrixCL& matr, const MatrixCL& other) {
    return matr.dot(other);
  }
  friend MatrixCL transpose(const MatrixCL& matr) {
    return matr.transpose();
  }
  friend MatrixCL operator+(T a, const MatrixCL& matr) {
    MatrixCL res(matr.nw_, matr.cl);
    matr.cl->Add(matr.nw_, matr.data_, a, res.data_);
    return res;
  }
  friend MatrixCL operator-(T a, const MatrixCL& matr) {
    MatrixCL res(matr.nw_, matr.cl);
    matr.cl->Sub(matr.nw_, a, matr.data_, res.data_);
    return res;
  }
  friend MatrixCL operator*(T a, const MatrixCL& matr) {
    MatrixCL res(matr.nw_, matr.cl);
    matr.cl->Mul(matr.nw_, matr.data_, a, res.data_);
    return res;
  }
  friend MatrixCL operator/(T a, const MatrixCL& matr) {
    MatrixCL res(matr.nw_, matr.cl);
    matr.cl->Div(matr.nw_, a, matr.data_, res.data_);
    return res;
  }
  friend MatrixCL roll(const MatrixCL& matr, int shift_row, int shift_col) {
    return matr.roll(matr.nw_, {shift_row, shift_col});
  }

  template <class U>
  friend MatrixCL conv(const MatrixCL& matr, const U& a, const U& axm,
                       const U& axp, const U& aym, const U& ayp) {
    return matr.conv(a, axm, axp, aym, ayp);
  }

  // Static functions.
  static MatrixCL zeros(size_t nrow, size_t ncol, CL& cl) {
    return MatrixCL(nrow, ncol, T(0), &cl);
  }
  static MatrixCL zeros(size_t n, CL& cl) {
    return MatrixCL::zeros(n, n, cl);
  }
  static MatrixCL ones(size_t nrow, size_t ncol, CL& cl) {
    return MatrixCL(nrow, ncol, T(1), &cl);
  }
  static MatrixCL ones(size_t n, CL& cl) {
    return MatrixCL::ones(n, n, cl);
  }
  template <class U>
  static MatrixCL zeros_like(const MatrixCL<U>& other) {
    return MatrixCL(other.nw_, T(0), other.cl);
  }
  template <class U>
  static MatrixCL ones_like(const MatrixCL<U>& other) {
    return MatrixCL(other.nw_, T(1), other.cl);
  }

 private:
  CL* cl = nullptr;
  MSize nw_;
  CL::Buffer<T> data_;
};

////////////////////////////////////////
// Output.
////////////////////////////////////////

template <class T>
std::ostream& operator<<(std::ostream& out, const MatrixCL<T>& matr) {
  out << Matrix<T>(matr);
  return out;
}

template <class T>
std::string MatrixToStr(const MatrixCL<T>& matr, int width = 3,
                        int precision = 6, bool fixed = false) {
  return MatrixToStr(Matrix<T>(matr), width, precision, fixed);
};
