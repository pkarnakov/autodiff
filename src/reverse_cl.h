#pragma once

#include "matrix_cl.h"
#include "reverse.h"

////////////////////////////////////////
// Specializations for reverse.h
////////////////////////////////////////
template <class T>
struct TypeName<MatrixCL<T>> {
  inline static const std::string value =
      "MatrixCL<" + TypeName<T>::value + ">";
};
template <class T>
void Clear(MatrixCL<T>& matr) {
  matr.clear();
};
template <class T, class E>
Tracer<T, E> sum(const Tracer<MatrixCL<T>, E>& tr_x) {
  return {std::make_shared<NodeUnary<T, MatrixCL<T>, E>>(
      tr_x.node(), [](const MatrixCL<T>& x) { return x.sum(); },
      [](const MatrixCL<T>& x, const T& du) {
        return du * MatrixCL<T>::ones_like(x);
      },
      "sum")};
}
template <class T, class E>
Tracer<T, E> mean(const Tracer<MatrixCL<T>, E>& tr_x) {
  return {std::make_shared<NodeUnary<T, MatrixCL<T>, E>>(
      tr_x.node(), [](const MatrixCL<T>& x) { return x.mean(); },
      [](const MatrixCL<T>& x, const T& du) {
        return du * MatrixCL<T>::ones_like(x) / x.size();
      },
      "mean")};
}
