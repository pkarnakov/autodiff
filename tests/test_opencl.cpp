#include <cstdio>
#include <fstream>
#include <iomanip>
#include <iostream>
#include <sstream>

#include "matrix_cl.h"
#include "opencl.h"
#include "reverse.h"
#include "reverse_cl.h"

// Print and evaluate.
#define PE(a) std::cout << #a << ": " << (a) << std::endl;
#define PEN(a) std::cout << #a << ":\n" << (a) << '\n' << std::endl;

using CL = OpenCL;

static CL Init(bool verbose = false) {
  CL::Config config;
  config.platform = 0;
  config.verbose = 0;
  CL cl(config);
  if (verbose) {
    std::cerr << "Device: " << cl.device_info_.name << std::endl;
    std::cerr << "Local size: " << cl.local_size_[0] << "," << cl.local_size_[1]
              << std::endl;
  }
  return cl;
}

template <class Scal = double>
static void TestBasic(CL& cl, size_t nrow) {
  std::cout << '\n' << __func__ << std::endl;
  auto& queue = cl.queue_;
  auto& context = cl.context_;

  const CL::MSize nw{nrow, nrow};
  CL::MirroredBuffer<Scal> u;
  CL::MirroredBuffer<Scal> v;
  u.Create(context, nw[0] * nw[1], CL_MEM_WRITE_ONLY);
  v.Create(context, nw[0] * nw[1], CL_MEM_WRITE_ONLY);
  for (size_t i = 0; i < u.size(); ++i) {
    u[i] = 3;
    v[i] = 10;
  }
  u.EnqueueWrite(queue);
  v.EnqueueWrite(queue);

  PE(cl.Sum(nw, u) / u.size());
  PE(cl.Dot(nw, u, v) / u.size());
}

template <class Scal = double>
static void TestMatrix(CL& cl, size_t nrow) {
  std::cout << '\n' << __func__ << std::endl;
  auto Str = [](auto m) { return MatrixToStr(m, 3, 3); };
  const size_t ncol = nrow;
  MatrixCL<Scal> u(nrow, ncol, cl);
  MatrixCL<Scal> v(nrow, ncol, cl);
  const MatrixCL<Scal> iota(Matrix<Scal>::iota(nrow, ncol), cl);
  u.fill(3);
  v.fill(10);
  PE(u.mean());
  PE(u.max());
  PE(u.min());
  PE(u.dot(v) / u.size());
  PE((u + v).mean());
  PE((u - v).mean());
  PE((u * v).mean());
  PE((u / v).mean());
  PE((u + 2).mean());
  PE((2 + u).mean());
  PE((u - 2).mean());
  PE((2 - u).mean());
  PE((u * 2).mean());
  PE((2 * u).mean());
  PE((sin(u)).mean());
  PE((cos(u)).mean());
  PE((exp(u)).mean());
  PE((log(u)).mean());
  PE(Matrix<Scal>(u).mean());
  PE(MatrixCL<Scal>(Matrix<Scal>(u) + 1, cl).mean());
  PE(u(2, 2));
  PE(u(2, 2) = 17);
  PE(u(2, 2));
  PE((u = v, u.mean()));
  PE((u += v, u.mean()));
  MatrixCL<Scal> w;
  w = u;
  PE((w = u, w.mean()));
  PEN(Str(iota.roll(0, 0)));
  PEN(Str(iota.roll(1, 0)));
  PEN(Str(iota.roll(0, 1)));
  PEN(Str(iota.roll(1, 1)));
  PEN(Str(iota.roll(-1, 0)));
  PEN(Str(iota.roll(0, -1)));
  PEN(Str(iota.roll(-1, -1)));

  u = iota;
  PEN(Str(u));
  PE(rms(u.conv(1, 0, 0, 0, 0) - u));
  PE(rms(u.conv(0, 1, 0, 0, 0) - u.roll(1, 0)));
  PE(rms(u.conv(0, 0, 1, 0, 0) - u.roll(-1, 0)));
  PE(rms(u.conv(0, 0, 0, 1, 0) - u.roll(0, 1)));
  PE(rms(u.conv(0, 0, 0, 0, 1) - u.roll(0, -1)));
  PEN(Str(u.conv(-4, 1, 1, 1, 1)));
}

template <class Scal = double>
static void TestMultigrid(CL& cl, size_t nrow) {
  std::cout << '\n' << __func__ << std::endl;
  auto Str = [](auto m) { return MatrixToStr(m, 3, 3); };
  const size_t ncol = nrow;
  MatrixCL<Scal> u(Matrix<Scal>::iota(nrow, ncol), cl);
  PEN(Str(u));
  PEN(Str(u.restrict()));
  PEN(Str(u.restrict().restrict_adjoint()));
  PEN(Str(u.restrict().interpolate()));
  PEN(Str(u.interpolate_adjoint()));
}

struct Extra : public BaseExtra {
  Extra(std::ostream& out) : dot(out) {}
  template <class Node>
  void Visit(Node* node) {
    dot.Write(node);
  }
  DotWriter dot;
};

template <class Scal = double>
static void TestReverse(CL& cl, size_t nrow) {
  std::cout << '\n' << __func__ << std::endl;
  const auto pi = M_PI;
  auto Str = [](auto m) { return MatrixToStr(m, 3, 1); };

  const size_t ncol = nrow;
  const auto eye = Matrix<Scal>::eye(nrow, ncol);
  auto matr_x = std::make_unique<MatrixCL<Scal>>(eye * (pi / 8), cl);
  auto matr_y = std::make_unique<MatrixCL<Scal>>(eye * (pi / 8), cl);
  Var<MatrixCL<Scal>> var_x(std::move(matr_x), "x");
  Var<MatrixCL<Scal>> var_y(std::move(matr_y), "y");
  PE(var_x);
  PE(var_y);
  auto x = MakeTracer<Extra>(var_x);
  auto y = MakeTracer<Extra>(var_y);

  auto eval = [&](auto& e, std::string path) {
    const auto order = e.GetFowardOrder();
    e.UpdateGrad(order);
    PE(e.value());
    PEN(Str(x.grad()));
    PEN(Str(y.grad()));
    std::cout << path << std::endl;
    std::ofstream fout(path);
    Extra extra(fout);
    Traverse(order, extra);
    ClearVisited(order);
  };
  auto e1 = sum(sin(x) * cos(y) + cos(x) * sin(y));
  eval(e1, "reverse_cl_1.gv");
  auto e2 = sum(sin(x + y));
  eval(e2, "reverse_cl_2.gv");
}

int main() {
  auto cl = Init(true);
  TestBasic(cl, 16);
  TestReverse(cl, 16);
  TestMatrix(cl, 4);
  TestMultigrid(cl, 8);
}
