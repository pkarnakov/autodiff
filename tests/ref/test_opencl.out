
TestBasic
cl.Sum(u) / u.size(): 3
cl.Dot(u, v): 30

TestReverse
var_x: Var(MatrixCL<double>, "x")
var_y: Var(MatrixCL<double>, "y")
e.value(): 11.3137
Str(x.grad()):
0.7   1   1   1   1   1   1   1   1   1   1   1   1   1   1   1
  1 0.7   1   1   1   1   1   1   1   1   1   1   1   1   1   1
  1   1 0.7   1   1   1   1   1   1   1   1   1   1   1   1   1
  1   1   1 0.7   1   1   1   1   1   1   1   1   1   1   1   1
  1   1   1   1 0.7   1   1   1   1   1   1   1   1   1   1   1
  1   1   1   1   1 0.7   1   1   1   1   1   1   1   1   1   1
  1   1   1   1   1   1 0.7   1   1   1   1   1   1   1   1   1
  1   1   1   1   1   1   1 0.7   1   1   1   1   1   1   1   1
  1   1   1   1   1   1   1   1 0.7   1   1   1   1   1   1   1
  1   1   1   1   1   1   1   1   1 0.7   1   1   1   1   1   1
  1   1   1   1   1   1   1   1   1   1 0.7   1   1   1   1   1
  1   1   1   1   1   1   1   1   1   1   1 0.7   1   1   1   1
  1   1   1   1   1   1   1   1   1   1   1   1 0.7   1   1   1
  1   1   1   1   1   1   1   1   1   1   1   1   1 0.7   1   1
  1   1   1   1   1   1   1   1   1   1   1   1   1   1 0.7   1
  1   1   1   1   1   1   1   1   1   1   1   1   1   1   1 0.7

Str(y.grad()):
0.7   1   1   1   1   1   1   1   1   1   1   1   1   1   1   1
  1 0.7   1   1   1   1   1   1   1   1   1   1   1   1   1   1
  1   1 0.7   1   1   1   1   1   1   1   1   1   1   1   1   1
  1   1   1 0.7   1   1   1   1   1   1   1   1   1   1   1   1
  1   1   1   1 0.7   1   1   1   1   1   1   1   1   1   1   1
  1   1   1   1   1 0.7   1   1   1   1   1   1   1   1   1   1
  1   1   1   1   1   1 0.7   1   1   1   1   1   1   1   1   1
  1   1   1   1   1   1   1 0.7   1   1   1   1   1   1   1   1
  1   1   1   1   1   1   1   1 0.7   1   1   1   1   1   1   1
  1   1   1   1   1   1   1   1   1 0.7   1   1   1   1   1   1
  1   1   1   1   1   1   1   1   1   1 0.7   1   1   1   1   1
  1   1   1   1   1   1   1   1   1   1   1 0.7   1   1   1   1
  1   1   1   1   1   1   1   1   1   1   1   1 0.7   1   1   1
  1   1   1   1   1   1   1   1   1   1   1   1   1 0.7   1   1
  1   1   1   1   1   1   1   1   1   1   1   1   1   1 0.7   1
  1   1   1   1   1   1   1   1   1   1   1   1   1   1   1 0.7

reverse_cl_1.gv
e.value(): 11.3137
Str(x.grad()):
0.7   1   1   1   1   1   1   1   1   1   1   1   1   1   1   1
  1 0.7   1   1   1   1   1   1   1   1   1   1   1   1   1   1
  1   1 0.7   1   1   1   1   1   1   1   1   1   1   1   1   1
  1   1   1 0.7   1   1   1   1   1   1   1   1   1   1   1   1
  1   1   1   1 0.7   1   1   1   1   1   1   1   1   1   1   1
  1   1   1   1   1 0.7   1   1   1   1   1   1   1   1   1   1
  1   1   1   1   1   1 0.7   1   1   1   1   1   1   1   1   1
  1   1   1   1   1   1   1 0.7   1   1   1   1   1   1   1   1
  1   1   1   1   1   1   1   1 0.7   1   1   1   1   1   1   1
  1   1   1   1   1   1   1   1   1 0.7   1   1   1   1   1   1
  1   1   1   1   1   1   1   1   1   1 0.7   1   1   1   1   1
  1   1   1   1   1   1   1   1   1   1   1 0.7   1   1   1   1
  1   1   1   1   1   1   1   1   1   1   1   1 0.7   1   1   1
  1   1   1   1   1   1   1   1   1   1   1   1   1 0.7   1   1
  1   1   1   1   1   1   1   1   1   1   1   1   1   1 0.7   1
  1   1   1   1   1   1   1   1   1   1   1   1   1   1   1 0.7

Str(y.grad()):
0.7   1   1   1   1   1   1   1   1   1   1   1   1   1   1   1
  1 0.7   1   1   1   1   1   1   1   1   1   1   1   1   1   1
  1   1 0.7   1   1   1   1   1   1   1   1   1   1   1   1   1
  1   1   1 0.7   1   1   1   1   1   1   1   1   1   1   1   1
  1   1   1   1 0.7   1   1   1   1   1   1   1   1   1   1   1
  1   1   1   1   1 0.7   1   1   1   1   1   1   1   1   1   1
  1   1   1   1   1   1 0.7   1   1   1   1   1   1   1   1   1
  1   1   1   1   1   1   1 0.7   1   1   1   1   1   1   1   1
  1   1   1   1   1   1   1   1 0.7   1   1   1   1   1   1   1
  1   1   1   1   1   1   1   1   1 0.7   1   1   1   1   1   1
  1   1   1   1   1   1   1   1   1   1 0.7   1   1   1   1   1
  1   1   1   1   1   1   1   1   1   1   1 0.7   1   1   1   1
  1   1   1   1   1   1   1   1   1   1   1   1 0.7   1   1   1
  1   1   1   1   1   1   1   1   1   1   1   1   1 0.7   1   1
  1   1   1   1   1   1   1   1   1   1   1   1   1   1 0.7   1
  1   1   1   1   1   1   1   1   1   1   1   1   1   1   1 0.7

reverse_cl_2.gv

TestMatrix
u.mean(): 3
u.max(): 3
u.min(): 3
u.dot(v) / u.size(): 30
(u + v).mean(): 13
(u - v).mean(): -7
(u * v).mean(): 30
(u / v).mean(): 0.3
(u + 2).mean(): 5
(2 + u).mean(): 5
(u - 2).mean(): 1
(2 - u).mean(): -1
(u * 2).mean(): 6
(2 * u).mean(): 6
(sin(u)).mean(): 0.14112
(cos(u)).mean(): -0.989992
(exp(u)).mean(): 20.0855
(log(u)).mean(): 1.09861
Matrix<Scal>(u).mean(): 3
MatrixCL<Scal>(Matrix<Scal>(u) + 1, cl).mean(): 4
u(2, 2): 3
u(2, 2) = 17: 17
u(2, 2): 17
(u = v, u.mean()): 10
(u += v, u.mean()): 20
(w = u, w.mean()): 20
Str(iota.roll(0, 0)):
  0   1   2   3
  4   5   6   7
  8   9  10  11
 12  13  14  15

Str(iota.roll(1, 0)):
 12  13  14  15
  0   1   2   3
  4   5   6   7
  8   9  10  11

Str(iota.roll(0, 1)):
  3   0   1   2
  7   4   5   6
 11   8   9  10
 15  12  13  14

Str(iota.roll(1, 1)):
 15  12  13  14
  3   0   1   2
  7   4   5   6
 11   8   9  10

Str(iota.roll(-1, 0)):
  4   5   6   7
  8   9  10  11
 12  13  14  15
  0   1   2   3

Str(iota.roll(0, -1)):
  1   2   3   0
  5   6   7   4
  9  10  11   8
 13  14  15  12

Str(iota.roll(-1, -1)):
  5   6   7   4
  9  10  11   8
 13  14  15  12
  1   2   3   0

Str(u):
  0   1   2   3
  4   5   6   7
  8   9  10  11
 12  13  14  15

rms(u.conv(1, 0, 0, 0, 0) - u): 0
rms(u.conv(0, 1, 0, 0, 0) - u.roll(1, 0)): 0
rms(u.conv(0, 0, 1, 0, 0) - u.roll(-1, 0)): 0
rms(u.conv(0, 0, 0, 1, 0) - u.roll(0, 1)): 0
rms(u.conv(0, 0, 0, 0, 1) - u.roll(0, -1)): 0
Str(u.conv(-4, 1, 1, 1, 1)):
 20  16  16  12
  4   0   0  -4
  4   0   0  -4
-12 -16 -16 -20

